Sphinx documentation (#194)
NCAR/DART: assimilation_code/programs/readme.rst

  Program to test some of the more fundamental routines in any ``model_mod``, especially a for a new model.

:doc:`perturb_single_instance <perturb_single_instance/perturb_single_instance>`
   Generate an ensemble of perturbed ensemble member restart files.
   (Alternatively, you might perturb the model state using ``model_nml`` variables).
 
:doc:`gen_sampling_err_table <gen_sampling_err_table/gen_sampling_err_table>`
   Computes a table of values needed to apply Sampling Error Correction (SEC),
   which corrects covariances based on small sample size statistics.
 
Creating Observation Sequence Files
-----------------------------------

obs_utils/create_obs_grid
   Create a set of observations located on a regular grid.  Obs have no data values.

:doc:`create_fixed_network_seq <create_fixed_network_seq/create_fixed_network_seq>` 
   Reads observation sequence file information from standard input 
   and replicates it multiple times in a second observation sequence file, at user specified dates. 
 
obs_utils/obs_timejitter
   Randomly perturb the times of the observations in a (usually) set_def.out file.
   Writes the results to (usually) obs_seq.in.

> :doc:`create_obs_sequence <create_obs_sequence/create_obs_sequence>`

Reviewer: suggested order in this section:  create_obs_sequence, create_fixed_network_seq, create_obs_grid obs_timejitter

Reviewer 2: My feeling is that when I go to a website, if I know what I'm looking for, I search for it and go directly there.
If I don't know, I want the pieces to be orgainized in some order that I can see or understand.
Ordering by how often a program is used or how frequently the documentation is referenced 
doesn't help me much.
Ordering by which programs create output which could be used as inputs by others makes sense to me.
I could be an outlier.

Reviewer: i agree. the order these programs are used is normally create_obs_sequence first, and output of that is the input to create_fixed_network_seq. the other two are infrequently used. so my suggested order in this section fits your criteria, does it not?

Reviewer 2: My mistake; I was led astray by the mention of obs_seq.in in create_obs_grid, 
but not in the create_obs_sequence.rst.  Now that I've read all the comments
it's clear that create_obs_sequence should be first.

Reviewer 2: Also, it seems that create_fixed_network_seq creates a time sequence from whatever array of obs locations
is fed to it, so create_obs_grid could or would create input for that.  But I'm gettig the impression 
that create_obs_grid was created just to create an obs_seq.in for perfect_model_obs.

   Prints out a quick table of obs types and counts, overall start and stop times, 
   and metadata strings and counts.  See obs_diag for more.
   There is an older version in the obs_assim_count directory.

:doc:`obs_seq_coverage <obs_seq_coverage/obs_seq_coverage>`
   Queries a set of observation sequence files to determine which observation locations report
   frequently enough to be useful for a verification study.
 
obs_total_error
   Prints the total error in the mean and spread from an obs_seq file 
   which has been through both perfect_model_obs and filter, so it has copies
   'truth', 'ensemble mean', and 'ensemble spread'.
   You can get more information by running the obs_diag program.

Changing Observation Sequence Files
-----------------------------------

:doc:`obs_keep_a_few <obs_keep_a_few/obs_keep_a_few>`
   Creates an output observation sequence file that is shorter than the input obs_seq file.
 
:doc:`obs_selection <obs_selection/obs_selection>`
   Extracts observations out of one or more obs_sequence files
   according to a  list of observation types, times, and locations.
   The list is usually created by :doc:`obs_seq_coverage/obs_seq_coverage`, 

>    but can be an observation sequence file,

Reviewer: is something missing at the end of this line?  it seems incomplete so i don't think the comma was intended to be a period.

Reviewer 2: Good catch!  The ',' should be a '.' (poor font choice in my editor)
and the link was missing punctuation and a link name.

> :doc:`obs_sequence_tool <obs_sequence_tool/obs_sequence_tool>`

Reviewer: suggested order for this section:  obs_sequence_tool, obs_common_subset, obs_selection, obs_seq_verify,  obs_loop, obs_keep_a_few, and then the remaining files.

Programs included with DART
===========================

This list of programs is separated into groups which have similar functionality.
Within each group they are sorted  by the order

> in which they might be used andor by how widely they are used.

Reviewer: ```suggestion
in which they might be used and/or by how widely they are used.
```

Setting Up Experiments
-----------------------------------

In many cases, you won't need to use any programs in Setting Up Experiments
except for ``preprocess``, because you're using an existing model interface 
and have the observation sequence files.
In that case, you the programs you're looking for are probably in `Assimilation Programs`_.
 

> `preprocess/preprocess <preprocess/preprocess.html>`_

Reviewer: ```suggestion
`preprocess <preprocess/preprocess.html>`_
```
for anything where the name of the executable and the subdirectory is the same, repeating the name just looks confusing.  for places where they differ, i'd keep the subdir name.

   Program to insert observation specific code into DART before filter or perfect_model_obs is compiled.
 

> `fill_inflation_restart/fill_inflation_restart <fill_inflation_restart/fill_inflation_restart.html>`_

Reviewer: ```suggestion
`fill_inflation_restart <fill_inflation_restart/fill_inflation_restart.html>`_
```

   Create inflation restart files with constant values taken from ``fill_inflation_restart_nml``.

> `obs_impact_tool/obs_impact_tool <obs_impact_tool/obs_impact_tool.html>`_

Reviewer: ```suggestion
`obs_impact_tool <obs_impact_tool/obs_impact_tool.html>`_
```

   Construct a table that is read by filter at run-time to localize the
   impact of sets of observation types on sets of state vector quantities.
 

> `model_mod_check/model_mod_check <model_mod_check/model_mod_check.html>`_

Reviewer: ```suggestion
`model_mod_check <model_mod_check/model_mod_check.html>`_
```

>    "Hackable" program to test some of the more fundamental routines in any ``model_mod``, 

Reviewer: i'm not sure what 'hackable' means.  
```suggestion
  Program to test some of the more fundamental routines in any ``model_mod``, 
```

   especially a new one.

> `perturb_single_instance/perturb_single_instance <perturb_single_instance/perturb_single_instance.html>`_

Reviewer: ```suggestion
`perturb_single_instance <perturb_single_instance/perturb_single_instance.html>`_
```

   Generate an ensemble of perturbed ensemble member restart files.
   (Alternatively, you might perturb the model state using ``model_nml`` variables).
 

> `gen_sampling_err_table/gen_sampling_err_table <gen_sampling_err_table/gen_sampling_err_table.html>`_

Reviewer: ```suggestion
`gen_sampling_err_table <gen_sampling_err_table/gen_sampling_err_table.html>`_
```

   Computes a table of values needed to apply Sampling Error Correction (SEC),
   which corrects covariances based on small sample size statistics.
 
Creating Observation Sequence Files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

obs_utils/create_obs_grid
   Create a set of observations located on a regular grid.  Obs have no data values.

> `create_fixed_network_seq/create_fixed_network_seq <create_fixed_network_seq/create_fixed_network_seq.html>`_ 

Reviewer: ```suggestion
`create_fixed_network_seq <create_fixed_network_seq/create_fixed_network_seq.html>`_ 
```

   Reads observation sequence file information from standard input 
   and replicates it multiple times in a second observation sequence file, at user specified dates. 
 
obs_utils/obs_timejitter
   Randomly perturb the times of the observations in a (usually) set_def.out file.
   Writes the results to (usually) obs_seq.in.

> `create_obs_sequence/create_obs_sequence <create_obs_sequence/create_obs_sequence.html>`_

Reviewer: ```suggestion
`create_obs_sequence <create_obs_sequence/create_obs_sequence.html>`_
```

   Creates a short andor synthetic observation sequence file using values read from standard input.
 
Querying Observation Sequence Files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

obs_utils/obs_info
   Summarize obs types, times, counts found in observation sequence file(s).

obs_utils/obs_assim_count
   Prints out a quick table of obs types and counts, overall start and stop times, 
   and metadata strings and counts.  See obs_diag for more.
   There is an older version in the obs_assim_count directory.

> `obs_seq_coverage/obs_seq_coverage <obs_seq_coverage/obs_seq_coverage.html>`_

Reviewer: ```suggestion
`obs_seq_coverage <obs_seq_coverage/obs_seq_coverage.html>`_
```

   Queries a set of observation sequence files to determine which observation locations report
   frequently enough to be useful for a verification study.
 

> obs_total_error/obs_total_error

Reviewer: ```suggestion
obs_total_error
```

   Prints the total error in the mean and spread from an obs_seq file 
   which has been through both perfect_model_obs and filter, so it has copies
   'truth', 'ensemble mean', and 'ensemble spread'.
   You can get more information by running the obs_diag program.

Changing Observation Sequence Files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> `obs_keep_a_few/obs_keep_a_few <obs_keep_a_few/obs_keep_a_few.html>`_

Reviewer: ```suggestion
`obs_keep_a_few <obs_keep_a_few/obs_keep_a_few.html>`_
```

   Creates an output observation sequence file that is shorter than the input obs_seq file.
 

> `obs_selection/obs_selection <obs_selection/obs_selection.html>`_

Reviewer: ```suggestion
`obs_selection <obs_selection/obs_selection.html>`_
```

   Extracts observations out of one or more obs_sequence files
   according to a  list of observation types, times, and locations.
   The list is usually created by :doc:`obs_seq_coverage/obs_seq_coverage`, 
   but can be an observation sequence file,
 

> `obs_sequence_tool/obs_sequence_tool <obs_sequence_tool/obs_sequence_tool.html>`_

Reviewer: since you said at the start of this file that the order inside the sections reflects the expected utility/use of the programs, i'd list this one first.  i'd move keep_a_few after obs_common_subset, and i'd put obs_selection, obs_coverage and obs_seq_verify together because they're sort of a matched set of executables.

Reviewer 3: incorporated all most all of that.

> `obs_sequence_tool/obs_sequence_tool <obs_sequence_tool/obs_sequence_tool.html>`_

Reviewer: ```suggestion
`obs_sequence_tool <obs_sequence_tool/obs_sequence_tool.html>`_
```
and move to start of this section.

   Subsets, combines, or alters observations from one or more observation sequence files 
   and optionally writes them into a single output obs_seq file.

> `obs_loop/obs_loop <obs_loop/obs_loop.html>`_

Reviewer: ```suggestion
`obs_loop <obs_loop/obs_loop.html>`_
```

   A template to read in observations from one obs_seq file and write them,
   optionally modified by user supplied code, to another obs_seq file.
 
obs_utils/obs_sort
   Do a complete sort of an obs_seq file by location, observation type, then variance.
   An ancestor of obs_remove_dups.

obs_utils/obs_remove_dups
   Removes duplicate observations from an obs_seq file, which involves a complete sort
   by time, location, observation type, then variance.
 

> `obs_common_subset/obs_common_subset <obs_common_subset/obs_common_subset.html>`_

Reviewer: ```suggestion
`obs_common_subset <obs_common_subset/obs_common_subset.html>`_
```

   Select the subset of observations, which were successfully assimilated, 
   from two or more assimilation cases (which used the same obs_seq.out file).
 

> `obs_seq_verify/obs_seq_verify <obs_seq_verify/obs_seq_verify.html>`_

Reviewer: ```suggestion
`obs_seq_verify <obs_seq_verify/obs_seq_verify.html>`_
```

   Reorders the observations from a forecast run of DART into a structure 
   that is amenable for the evaluation of the forecast.
 

obs_utils/obs_data_denial
   THIS IS NOT YET DONE!
   Help implement a data-denial experiment by randomly changing the error variance
   of N of each obs type in an observation sequence file to a huge value.
 
Assimilation Programs
-----------------------------------
 

> `perfect_model_obs/perfect_model_obs <perfect_model_obs/perfect_model_obs.html>`_

Reviewer: ```suggestion
`perfect_model_obs <perfect_model_obs/perfect_model_obs.html>`_
```

   Creates synthetic observation sequences from a hindcast model.
 

> `filter/filter <filter/filter.html>`_

Reviewer: ```suggestion
`filter <filter/filter.html>`_
```

   Main Fortran program for driving ensemble filter assimilations.

> filter/filter.separate_seq

Reviewer: remove this?

Author: done

   Like filter, but each task updates its own sequence in obs_space_diagnostics.
   Included here just for future reference.

> `advance_time/advance_time <advance_time/advance_time.html>`_

Reviewer: ```suggestion
`advance_time <advance_time/advance_time.html>`_
```

   Provides a shell-scripting-friendly way to increment and decrement calendar dates and times.
 

> `integrate_model/integrate_model <integrate_model/integrate_model.html>`_

Reviewer: ```suggestion
`integrate_model <integrate_model/integrate_model.html>`_
```

   Generic main program which advances a single ensemble member in ``perfect_model_obs`` 
   or the serial or parallel version of the ``filter`` program.

> `integrate_model/integrate_model_parallel <integrate_model/integrate_model.html>`_

Reviewer: combine these, or remove the parallel version altoghterh

Author: removed parallel version (not used anywhere it seems)

   Generic main program which advances a single
   ensemble member in ``perfect_model_obs`` or the serial ``
   or parallel version of the ``filter`` program.``

Evaluating Results
-----------------------------------
 
obs_diag 
   Reads obs_seq.final files, calculates statistics, and writes them to NetCDF files 
   for use by Matlab (or other) plotting scripts.

>    There are separate versions for models with different coordinate systems:
>    `1D <obs_diag/oned/obs_diag.html>`_,
>    `3D cartesian <obs_diag/threed_cartesian/obs_diag.html>`_,
>    `3D spherical <obs_diag/threed_sphere/obs_diag.html>`_, and 
>    3D spherical with streamflow.

Reviewer: ```suggestion
   There are separate versions for models with different coordinate systems:
   
   - `1D <obs_diag/oned/obs_diag.html>`_,
   - `3D cartesian <obs_diag/threed_cartesian/obs_diag.html>`_,
   - `3D spherical <obs_diag/threed_sphere/obs_diag.html>`_, and 
   - 3D spherical with streamflow.
```

or whatever formatting makes this into a list instead of putting them on the same line?  easier to see that way.

Author: done also capital Cartesian.

   
obs_seq_to_netcdf
   Extracts the observation components from observation sequence files and writes out
   netCDF files that can be used by other applications.
   such as ``diagnostics/matlab/plot_obs_netcdf*``

>    There are two versions; the is `standard version <obs_seq_to_netcdf/obs_seq_to_netcdf.html>`_

Reviewer: ```suggestion
   There are two versions; the `standard version <obs_seq_to_netcdf/obs_seq_to_netcdf.html>`_
```
(grammar)

   and one which filters out radiance metadata which is not needed by the scripts 
   which use the resulting NetCDF file.

> `compare_states/compare_states <compare_states/compare_states.html>`_

Reviewer: ```suggestion
`compare_states <compare_states/compare_states.html>`_
```

   Compare fields in two NetCDF files and print out the min and max values from each file and of
   the difference between the two files.

> `compute_error/compute_error <compute_error/compute_error.html>`_

Reviewer: ```suggestion
`compute_error <compute_error/compute_error.html>`_
```

   Compute the time-mean ensemble error and spread in the same manner as the DART MATLAB diagnostic
   routine ``plot_total_err``; in state space from true_state.nc and preassim.nc (or analysis.nc).
 

> `closest_member_tool/closest_member_tool <closest_member_tool/closest_member_tool.html>`_
>    Prints out a sorted order of which ensemble members are 'closest' to the mean, 
>    where 'close' is selectable by namelist option.

Reviewer: ```suggestion
`closest_member_tool <closest_member_tool/closest_member_tool.html>`_
   Prints out a sorted order of which ensemble members are 'closest' to the mean, 
   where the method for computing the 'close' metric is selectable by namelist option.
```

 
Historical and Deprecated
-------------------------
 
`system_simulation <system_simulation/system_simulation.html>`_
   A collection of standalone programs for simulating various properties of ensembles.
   Talk to Jeff Anderson about the programs in this directory.

system_simulation/system_simulation
   This program begins attempts to analyze the value of particular 
   observations. Begin by trying to determine the value of 
   observations with a given correlation to a state variable using an 

>    N member ensemble to compute the correlations.

Reviewer: ths seems redundant - remove?

Author: done

> `restart_file_tool/restart_file_tool <restart_file_tool/restart_file_tool.html>`_

Reviewer: ```suggestion
`restart_file_tool <restart_file_tool/restart_file_tool.html>`_
```

   Deprecated, since in Manhattan all DART initial and restart files are in NetCDF format.
 

> `wakeup_filter/wakeup_filter <wakeup_filter/wakeup_filter.html>`_

Reviewer: ```suggestion
`wakeup_filter <wakeup_filter/wakeup_filter.html>`_
```
